File: 0b-kernel/track-bilyy-garaev-mukhin/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"sort"
	"strconv"
	"strings"
	"text/tabwriter"
)

type FileStats struct {
	Path               string
	AccumLOC           int
	NetLOC             int
	NetLOCRootMid      int
	NetLOCMidHead      int
	SizeFirst          int64
	SizeMid            int64
	SizeLast           int64
	BinaryDelta        int64
	BinaryDeltaRootMid int64
	BinaryDeltaMidHead int64
}

type DirStats struct {
	Dir         string
	Files       int
	AccumLOC    int64
	NetLOC      int64
	BinaryDelta int64
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "usage: %s <repo-path>\n", os.Args[0])
		os.Exit(1)
	}
	repo := os.Args[1]

	fmt.Fprintln(os.Stderr, "finding root...")
	rootOut, err := runGit(repo, "rev-list", "--max-parents=0", "HEAD")
	if err != nil {
		fatal("root", err)
	}
	roots := strings.Fields(rootOut)
	if len(roots) == 0 {
		fatal("no root", fmt.Errorf("empty"))
	}
	root := roots[0]
	if len(root) > 12 {
		root = root[:12]
	}
	fmt.Fprintln(os.Stderr, "root:", root)

	mid := getMiddleCommit(repo, roots[0])
	if mid != "" {
		fmt.Fprintln(os.Stderr, "mid:", mid)
	}

	fmt.Fprintln(os.Stderr, "accum loc...")
	accum := collectAccum(repo)
	fmt.Fprintln(os.Stderr, "net loc root -> HEAD...")
	net := collectNetBetween(repo, roots[0], "HEAD")
	var netRootMid, netMidHead map[string]int
	if mid != "" {
		fmt.Fprintln(os.Stderr, "net loc root -> mid, mid -> HEAD...")
		netRootMid = collectNetBetween(repo, roots[0], mid)
		netMidHead = collectNetBetween(repo, mid, "HEAD")
	}
	fmt.Fprintln(os.Stderr, "sizes...")
	headSizes := collectSizes(repo, "HEAD")
	rootSizes := collectSizes(repo, roots[0])
	var midSizes map[string]int64
	if mid != "" {
		midSizes = collectSizes(repo, mid)
	}

	stats := merge(accum, net, netRootMid, netMidHead, headSizes, rootSizes, midSizes)
	sort.Slice(stats, func(i, j int) bool { return stats[i].AccumLOC > stats[j].AccumLOC })

	printTop(stats, 30)
	printDirs(stats)
	printDeltaVsLines(stats, 25)
	printSummary(stats)
}

func collectAccum(repo string) map[string]int {
	cmd := exec.Command("git", "log", "--numstat", "--format=", "--no-renames")
	cmd.Dir = repo
	out, _ := cmd.StdoutPipe()
	cmd.Start()

	m := make(map[string]int)
	sc := bufio.NewScanner(out)
	buf := make([]byte, 0, 512*1024)
	sc.Buffer(buf, 1024*1024)
	for sc.Scan() {
		line := sc.Text()
		if line == "" {
			continue
		}
		add, del, path, ok := parseNumstat(line)
		if !ok {
			continue
		}
		m[path] += add + del
	}
	cmd.Wait()
	return m
}

func getMiddleCommit(repo, root string) string {
	out, err := runGit(repo, "rev-list", "--first-parent", root+"..HEAD")
	if err != nil || out == "" {
		return ""
	}
	commits := strings.Split(strings.TrimSpace(out), "\n")
	if len(commits) < 2 {
		return ""
	}
	mid := commits[len(commits)/2]
	if len(mid) > 12 {
		mid = mid[:12]
	}
	return mid
}

func collectNetBetween(repo, from, to string) map[string]int {
	cmd := exec.Command("git", "diff", "--numstat", "--no-renames", from, to)
	cmd.Dir = repo
	out, _ := cmd.StdoutPipe()
	cmd.Start()

	m := make(map[string]int)
	sc := bufio.NewScanner(out)
	for sc.Scan() {
		add, del, path, ok := parseNumstat(sc.Text())
		if !ok {
			continue
		}
		m[path] = add + del
	}
	cmd.Wait()
	return m
}

func collectSizes(repo, ref string) map[string]int64 {
	out, err := runGit(repo, "ls-tree", "-r", "-l", ref)
	if err != nil {
		fatal("ls-tree", err)
	}
	m := make(map[string]int64)
	sc := bufio.NewScanner(strings.NewReader(out))
	for sc.Scan() {
		line := sc.Text()
		i := strings.IndexByte(line, '\t')
		if i < 0 {
			continue
		}
		path := line[i+1:]
		f := strings.Fields(line[:i])
		if len(f) < 4 {
			continue
		}
		size, err := strconv.ParseInt(f[3], 10, 64)
		if err != nil {
			continue
		}
		m[path] = size
	}
	return m
}

func merge(accum, net map[string]int, netRootMid, netMidHead map[string]int, headS, rootS map[string]int64, midS map[string]int64) []FileStats {
	var list []FileStats
	for path, lastSize := range headS {
		firstSize := rootS[path]
		delta := lastSize - firstSize
		if delta < 0 {
			delta = -delta
		}
		midSize := int64(0)
		deltaRootMid := int64(0)
		deltaMidHead := int64(0)
		if midS != nil {
			midSize = midS[path]
			d1 := midSize - firstSize
			if d1 < 0 {
				d1 = -d1
			}
			deltaRootMid = d1
			d2 := lastSize - midSize
			if d2 < 0 {
				d2 = -d2
			}
			deltaMidHead = d2
		}
		list = append(list, FileStats{
			Path:               path,
			AccumLOC:           accum[path],
			NetLOC:             net[path],
			NetLOCRootMid:      at(netRootMid, path),
			NetLOCMidHead:      at(netMidHead, path),
			SizeFirst:          firstSize,
			SizeMid:            midSize,
			SizeLast:           lastSize,
			BinaryDelta:        delta,
			BinaryDeltaRootMid: deltaRootMid,
			BinaryDeltaMidHead: deltaMidHead,
		})
	}
	return list
}

func at(m map[string]int, key string) int {
	if m == nil {
		return 0
	}
	return m[key]
}

func printTop(stats []FileStats, n int) {
	if n > len(stats) {
		n = len(stats)
	}
	fmt.Println()
	fmt.Println("--- top files by accum loc ---")
	tw := tabwriter.NewWriter(os.Stdout, 0, 4, 2, ' ', 0)
	fmt.Fprintln(tw, "file\taccum\tnet\tratio\tsize_first\tsize_mid\tsize_last\tdelta\tdelta_mid\tbytes/line")
	for i := 0; i < n; i++ {
		s := stats[i]
		path := s.Path
		if len(path) > 50 {
			path = "..." + path[len(path)-47:]
		}
		fmt.Fprintf(tw, "%s\t%d\t%d\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n",
			path, s.AccumLOC, s.NetLOC, ratio(s.AccumLOC, s.NetLOC),
			bytesFmt(s.SizeFirst), bytesFmt(s.SizeMid), bytesFmt(s.SizeLast),
			bytesFmt(s.BinaryDelta), bytesFmt(s.BinaryDeltaMidHead),
			bytesPerLine(s.BinaryDelta, s.NetLOC))
	}
	tw.Flush()
}

func bytesPerLine(deltaBytes int64, netLOC int) string {
	if netLOC <= 0 {
		return "-"
	}
	return fmt.Sprintf("%.1f", float64(deltaBytes)/float64(netLOC))
}

func printDeltaVsLines(stats []FileStats, n int) {
	var withLOC []FileStats
	for _, s := range stats {
		if s.NetLOC > 0 {
			withLOC = append(withLOC, s)
		}
	}
	sort.Slice(withLOC, func(i, j int) bool {
		bi := float64(withLOC[i].BinaryDelta) / float64(withLOC[i].NetLOC)
		bj := float64(withLOC[j].BinaryDelta) / float64(withLOC[j].NetLOC)
		return bi > bj
	})
	if n > len(withLOC) {
		n = len(withLOC)
	}
	fmt.Println()
	fmt.Println("--- per-file: byte delta vs net LOC (bytes per line, top by B/L) ---")
	tw := tabwriter.NewWriter(os.Stdout, 0, 4, 2, ' ', 0)
	fmt.Fprintln(tw, "file\tnet_LOC\tdelta_bytes\tbytes/line")
	for i := 0; i < n; i++ {
		s := withLOC[i]
		path := s.Path
		if len(path) > 50 {
			path = "..." + path[len(path)-47:]
		}
		bpl := float64(s.BinaryDelta) / float64(s.NetLOC)
		fmt.Fprintf(tw, "%s\t%d\t%s\t%.1f\n", path, s.NetLOC, bytesFmt(s.BinaryDelta), bpl)
	}
	tw.Flush()
}

func printDirs(stats []FileStats) {
	byDir := make(map[string]*DirStats)
	for _, s := range stats {
		d := topDir(s.Path)
		if byDir[d] == nil {
			byDir[d] = &DirStats{Dir: d}
		}
		byDir[d].Files++
		byDir[d].AccumLOC += int64(s.AccumLOC)
		byDir[d].NetLOC += int64(s.NetLOC)
		byDir[d].BinaryDelta += s.BinaryDelta
	}
	var dirs []*DirStats
	for _, ds := range byDir {
		dirs = append(dirs, ds)
	}
	sort.Slice(dirs, func(i, j int) bool { return dirs[i].AccumLOC > dirs[j].AccumLOC })

	fmt.Println()
	fmt.Println("--- by directory ---")
	tw := tabwriter.NewWriter(os.Stdout, 0, 4, 2, ' ', 0)
	fmt.Fprintln(tw, "dir\tfiles\taccum\tnet\tratio\tdelta")
	for _, ds := range dirs {
		fmt.Fprintf(tw, "%s\t%d\t%d\t%d\t%s\t%s\n",
			ds.Dir, ds.Files, ds.AccumLOC, ds.NetLOC,
			ratio(int(ds.AccumLOC), int(ds.NetLOC)), bytesFmt(ds.BinaryDelta))
	}
	tw.Flush()
}

func printSummary(stats []FileStats) {
	var totalAccum, totalNet int64
	var totalDelta, totalDeltaRootMid, totalDeltaMidHead int64
	var totalNetRootMid, totalNetMidHead int64
	maxRatio := 0.0
	maxFile := ""

	for _, s := range stats {
		totalAccum += int64(s.AccumLOC)
		totalNet += int64(s.NetLOC)
		totalDelta += s.BinaryDelta
		totalDeltaRootMid += s.BinaryDeltaRootMid
		totalDeltaMidHead += s.BinaryDeltaMidHead
		totalNetRootMid += int64(s.NetLOCRootMid)
		totalNetMidHead += int64(s.NetLOCMidHead)
		if s.NetLOC > 0 {
			r := float64(s.AccumLOC) / float64(s.NetLOC)
			if r > maxRatio {
				maxRatio = r
				maxFile = s.Path
			}
		}
	}

	fmt.Println()
	fmt.Println("--- summary ---")
	fmt.Printf("files: %d\n", len(stats))
	fmt.Printf("accum loc: %d\n", totalAccum)
	fmt.Printf("net loc (root -> HEAD): %d\n", totalNet)
	fmt.Printf("ratio: %s\n", ratio(int(totalAccum), int(totalNet)))
	fmt.Printf("binary delta (root -> HEAD): %s\n", bytesFmt(totalDelta))
	if totalNetRootMid > 0 || totalDeltaRootMid > 0 {
		fmt.Println()
		fmt.Println("--- middle point (root -> mid, mid -> HEAD) ---")
		fmt.Printf("net loc root -> mid: %d\n", totalNetRootMid)
		fmt.Printf("net loc mid -> HEAD: %d\n", totalNetMidHead)
		fmt.Printf("binary delta root -> mid: %s\n", bytesFmt(totalDeltaRootMid))
		fmt.Printf("binary delta mid -> HEAD: %s\n", bytesFmt(totalDeltaMidHead))
		if totalNetRootMid > 0 {
			fmt.Printf("bytes/line root -> mid: %.1f\n", float64(totalDeltaRootMid)/float64(totalNetRootMid))
		}
		if totalNetMidHead > 0 {
			fmt.Printf("bytes/line mid -> HEAD: %.1f\n", float64(totalDeltaMidHead)/float64(totalNetMidHead))
		}
	}
	if maxFile != "" {
		fmt.Printf("max ratio file: %s (%.1fx)\n", maxFile, maxRatio)
	}
}

func runGit(repo string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = repo
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return string(out), nil
}

func parseNumstat(line string) (add, del int, path string, ok bool) {
	parts := strings.SplitN(line, "\t", 3)
	if len(parts) != 3 {
		return 0, 0, "", false
	}
	if parts[0] == "-" || parts[1] == "-" {
		return 0, 0, "", false
	}
	a, e1 := strconv.Atoi(parts[0])
	d, e2 := strconv.Atoi(parts[1])
	if e1 != nil || e2 != nil {
		return 0, 0, "", false
	}
	return a, d, parts[2], true
}

func topDir(path string) string {
	i := strings.IndexByte(path, '/')
	if i < 0 {
		return path
	}
	return path[:i]
}

func ratio(a, b int) string {
	if b == 0 {
		if a == 0 {
			return "-"
		}
		return "inf"
	}
	return fmt.Sprintf("%.2fx", float64(a)/float64(b))
}

func bytesFmt(n int64) string {
	if n >= 1<<30 {
		return fmt.Sprintf("%.1fG", float64(n)/(1<<30))
	}
	if n >= 1<<20 {
		return fmt.Sprintf("%.1fM", float64(n)/(1<<20))
	}
	if n >= 1<<10 {
		return fmt.Sprintf("%.1fK", float64(n)/(1<<10))
	}
	return fmt.Sprintf("%d", n)
}

func fatal(msg string, err error) {
	fmt.Fprintf(os.Stderr, "error %s: %v\n", msg, err)
	os.Exit(1)
}

File: 0b-kernel/track-bilyy-garaev-mukhin/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <ctype.h>

static void die(const char *msg) {
    perror(msg);
    exit(1);
}

static char *xstrdup(const char *s) {
    size_t n = strlen(s) + 1;
    char *p = malloc(n);
    if (!p) die("malloc");
    memcpy(p, s, n);
    return p;
}

typedef struct {
    char    *key;
    int64_t  value;
    int      used;
} MapEntry;

typedef struct {
    MapEntry *entries;
    size_t    cap;
    size_t    len;
} HashMap;

static uint64_t hash_str(const char *s) {
    uint64_t h = 1469598103934665603ULL;
    for (; *s; s++) {
        h ^= (unsigned char)*s;
        h *= 1099511628211ULL;
    }
    return h;
}

static void map_init(HashMap *m, size_t cap) {
    m->cap = cap;
    m->len = 0;
    m->entries = calloc(cap, sizeof(MapEntry));
    if (!m->entries) die("calloc");
}

static void map_grow(HashMap *m);

static MapEntry *map_find(HashMap *m, const char *key) {
    if (m->len * 2 >= m->cap)
        map_grow(m);

    uint64_t h = hash_str(key);
    size_t i = h % m->cap;

    for (;;) {
        MapEntry *e = &m->entries[i];
        if (!e->used)
            return e;
        if (strcmp(e->key, key) == 0)
            return e;
        i = (i + 1) % m->cap;
    }
}

static void map_add(HashMap *m, const char *key, int64_t delta) {
    MapEntry *e = map_find(m, key);
    if (!e->used) {
        e->used = 1;
        e->key = xstrdup(key);
        e->value = delta;
        m->len++;
    } else {
        e->value += delta;
    }
}

static void map_set(HashMap *m, const char *key, int64_t value) {
    MapEntry *e = map_find(m, key);
    if (!e->used) {
        e->used = 1;
        e->key = xstrdup(key);
        e->value = value;
        m->len++;
    } else {
        e->value = value;
    }
}

static int map_get(HashMap *m, const char *key, int64_t *out) {
    uint64_t h = hash_str(key);
    size_t i = h % m->cap;

    for (;;) {
        MapEntry *e = &m->entries[i];
        if (!e->used)
            return 0;
        if (strcmp(e->key, key) == 0) {
            *out = e->value;
            return 1;
        }
        i = (i + 1) % m->cap;
    }
}

static void map_grow(HashMap *m) {
    HashMap n;
    map_init(&n, m->cap * 2);

    for (size_t i = 0; i < m->cap; i++) {
        MapEntry *e = &m->entries[i];
        if (e->used)
            map_set(&n, e->key, e->value);
    }

    free(m->entries);
    *m = n;
}

static void map_free(HashMap *m) {
    for (size_t i = 0; i < m->cap; i++) {
        if (m->entries[i].used)
            free(m->entries[i].key);
    }
    free(m->entries);
}

static FILE *git_popen(const char *repo, const char *cmd) {
    char buf[4096];
    snprintf(buf, sizeof(buf), "cd '%s' && %s", repo, cmd);
    FILE *fp = popen(buf, "r");
    if (!fp) die("popen");
    return fp;
}

static void git_pclose(FILE *fp) {
    if (pclose(fp) == -1)
        die("pclose");
}

static void collect_accumulated_loc(const char *repo, HashMap *map) {
    FILE *fp = git_popen(
        repo,
        "git log --numstat --format= --no-renames"
    );

    char line[8192];
    while (fgets(line, sizeof(line), fp)) {
        char *a = strtok(line, "\t");
        char *d = strtok(NULL, "\t");
        char *p = strtok(NULL, "\n");

        if (!a || !d || !p)
            continue;
        if (a[0] == '-' || d[0] == '-')
            continue;

        int added = atoi(a);
        int deleted = atoi(d);
        map_add(map, p, added + deleted);
    }

    git_pclose(fp);
}

static void collect_net_loc(const char *repo, const char *root, HashMap *map) {
    char cmd[512];
    snprintf(cmd, sizeof(cmd),
        "git diff --numstat %s HEAD --no-renames", root);

    FILE *fp = git_popen(repo, cmd);

    char line[8192];
    while (fgets(line, sizeof(line), fp)) {
        char *a = strtok(line, "\t");
        char *d = strtok(NULL, "\t");
        char *p = strtok(NULL, "\n");

        if (!a || !d || !p)
            continue;
        if (a[0] == '-' || d[0] == '-')
            continue;

        int added = atoi(a);
        int deleted = atoi(d);
        map_set(map, p, added - deleted);
    }

    git_pclose(fp);
}

static void collect_sizes(const char *repo, const char *ref, HashMap *map) {
    char cmd[256];
    snprintf(cmd, sizeof(cmd),
        "git ls-tree -r -l %s", ref);

    FILE *fp = git_popen(repo, cmd);
    char line[8192];

    while (fgets(line, sizeof(line), fp)) {
        char *tab = strchr(line, '\t');
        if (!tab)
            continue;

        long size = 0;
        if (sscanf(line, "%*s %*s %*s %ld", &size) != 1)
            continue;

        char *path = tab + 1;
        path[strcspn(path, "\n")] = 0;

        map_set(map, path, size);
    }

    git_pclose(fp);
}

typedef struct {
    char   *path;
    int64_t accum_loc;
    int64_t net_loc;
    int64_t size_first;
    int64_t size_last;
    int64_t bin_delta;
} FileStat;

static int cmp_accum_desc(const void *a, const void *b) {
    const FileStat *x = a;
    const FileStat *y = b;
    return (y->accum_loc > x->accum_loc) - (y->accum_loc < x->accum_loc);
}

int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "usage: %s <git-repo>\n", argv[0]);
        return 1;
    }

    const char *repo = argv[1];

    FILE *fp = git_popen(repo, "git rev-list --max-parents=0 HEAD");
    char root[64];
    if (!fgets(root, sizeof(root), fp))
        die("cannot get root commit");
    root[strcspn(root, "\n")] = 0;
    git_pclose(fp);

    HashMap accum, net, size0, size1;
    map_init(&accum, 1 << 15);
    map_init(&net,   1 << 15);
    map_init(&size0, 1 << 15);
    map_init(&size1, 1 << 15);

    collect_accumulated_loc(repo, &accum);
    collect_net_loc(repo, root, &net);
    collect_sizes(repo, root, &size0);
    collect_sizes(repo, "HEAD", &size1);

    FileStat *stats = calloc(accum.len, sizeof(FileStat));
    if (!stats) die("calloc");

    size_t n = 0;
    for (size_t i = 0; i < accum.cap; i++) {
        MapEntry *e = &accum.entries[i];
        if (!e->used)
            continue;

        FileStat *s = &stats[n++];
        s->path = e->key;
        s->accum_loc = e->value;

        map_get(&net,   s->path, &s->net_loc);
        map_get(&size0, s->path, &s->size_first);
        map_get(&size1, s->path, &s->size_last);

        s->bin_delta = s->size_last - s->size_first;
    }

    qsort(stats, n, sizeof(FileStat), cmp_accum_desc);

    printf("%-60s %12s %12s %12s\n",
           "file", "accum_loc", "net_loc", "bin_delta");

    for (size_t i = 0; i < n; i++) {
        FileStat *s = &stats[i];
        printf("%-60s %12lld %12lld %12lld\n",
               s->path,
               (long long)s->accum_loc,
               (long long)s->net_loc,
               (long long)s->bin_delta);
    }

    free(stats);
    map_free(&accum);
    map_free(&net);
    map_free(&size0);
    map_free(&size1);

    return 0;
}
